test/software/amazon/ion/impl/BaseSymbolTableWrapperTest.java

```diff
@@ -0,0 +1,260 @@
+package software.amazon.ion.impl;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+import static org.mockito.Mockito.verifyZeroInteractions;
+import org.mockito.runners.MockitoJUnitRunner;
+import software.amazon.ion.IonWriter;
+import software.amazon.ion.SymbolTable;
+
+@RunWith(MockitoJUnitRunner.class)
```

Comment:
The purpose of this test is to show what is the expected contract for `BaseSymbolTableWrapper` which is to delegate all non overridden methods to the delegate. 

The mock here is used not to replace a pkg external dependency, but to validate the contract independent of any particular implementation

I don't agree this class is so trivial it almost doesn't need tests. Tests are not only to prove correctness, they are also a live documentation for classes/methods contracts. IMHO the contract documenting side of it is more important for abstract classes as it informs the developer what's to be expected when they subclass it 

Not using Mockito here would require either a fake impl that verifies the delegate methods were called or an impl with baked in responses to check the delegation behavior indirectly. Both seem more work and more convoluted than to use a mock framework which is designed for this type of case

I'm assuming that by "lasting implications" you mean opening the door to overuse mockito and create brittle tests that are too tied to the impl instead of testing the class/method contract. If that's the case I agree that we should avoid Mockito misuse. Being mindful in code reviews, having a best practices guide, documenting in the code the mock intention, which I should add to my change, are better ways of doing it than not having the dependency when it's useful for some types of tests

Reply:
I'd argue that since `BaseSymbolTableWrapper` is abstract, we shouldn't be testing it directly, but should rather be testing its implementations (namely, `LocalSymbolTableImportAdapter`). `BaseSymbolTableWrapperTest` tests an anonymous implementation, which will never exist in non-test code.

I understand documenting the contract, but that's already done in the class-level Javadoc. This is an internal class, so modifications that break the documented contract should be caught in review (in addition to being caught by the unit tests for the implementing classes).

When I talk about lasting implications, part of it is as you say: mocking makes it really easy to write tests that either don't test anything interesting, or test the wrong thing. I find mocking frameworks most useful when they are used to mock out external dependencies, but ion-java doesn't have any of those. As such, we've been able to avoid relying on a mocking framework throughout ion-java's 10 year development history.

If you think that testing `BaseSymbolTableWrapper` is necessary, I think we agree that a mocking framework may be convenient, but isn't required. I don't think that it actually requires much more work to avoid using one (and is likely easier to read for a mocking novice). A fake `SymbolTable` implementation would be required to verify calling behavior, but although this requires more lines of code, it's trivial to write. The test cases themselves would probably occupy about the same number of lines.

If this is a minority viewpoint, I'm happy to drop it and move on; it might be worthwhile to try to understand whether the ion-java developers who preceded us had a rationale behind the lack of a mocking framework.